_posts/2022/2022-01-06-github-actions-a-basic-how-to.md

```diff
@@ -0,0 +1,152 @@
+
+Let's have a look at some reasons why GitHub Actions could be a good choice:
+- The price is competitive ($0.008 per Linux minutes as of Dec/2021)
+- It fully integrates into GitHub (if you use it for hosting your repos)
+- YAML config → this means changes can be reviewed
+- It has a marketplace with many ready-to-use actions (e.g. triggering a webhook)
+
+## Build pipeline
+
+Great, so GitHub Actions has many advantages.
+But let's focus on one of its main use cases: creating your own build pipeline.
+You can build, test, and even deploy your service via GitHub Actions.
+You just need to know which commands to run.
+
+### How we do it
+
+In the case of our services, most of them are running with Gradle, so building them is really a no-brainer.
+Here are the steps needed for the build process of main branch builds (the ones which get deployed):
+- Compile and run unit/integration tests (`./gradlew build`)
+- Build and push a Docker image
+- Build and publish a Helm chart
+
+Quite straightforward, nothing fancy.
+We have set up GitHub Actions to run the build process in two different ocasions:
+- On pull requests (run tests)
+- When a pull request is merged into the main branch
+
+### Down to the code
+
+Here is an adapted version of our build’s YAML file (located within the `.github/workflows` directory):
+```yaml
+name: Build and test
+
+on:
+  push:                                 #######
+    branches: [ main, '*-test' ]        ## 1 ##
+  pull_request:                         #######
+    branches: [ '*' ]
+
+concurrency:                            #######
+  group: ${{ github.ref }}              ## 2 ##
+  cancel-in-progress: true              #######
+
+env:
+  # Gradle variables                                                              #######
+  ORG_GRADLE_PROJECT_dockerHubUsername: ${{ secrets.DOCKERHUB_USER }}             ## 3 ##
+  ORG_GRADLE_PROJECT_dockerHubPassword: ${{ secrets.DOCKERHUB_PASSWORD }}         #######
+
+jobs:
+  build:
+    name: Build / test / docker image 🐳
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2                               ### 4
+      - name: Set up JDK 11                                     ### 5
+        uses: actions/setup-java@v2
+        with:
+          java-version: 11
+          distribution: 'temurin' # adopt openjdk
+          cache: gradle
+      - name: Gradle test                                       ### 6
+        run: ./gradlew --no-daemon check asciidoctor
+
+      ### THE FOLLOWING STEPS ARE FOR MASTER BUILDS ONLY
+
+      - name: Login to Docker Hub
+        if: github.ref == 'refs/heads/master'
+        uses: docker/login-action@v1                                                #######
+        with:                                                                       ## 7 ##
+          username: ${{ secrets.BYD_DOCKERHUB_USER }}                               #######
```

Comment:
Somehow, parts starting with $ are not included in the rendered version. I'm on it and checked that back with Team 42. @roberto-filho would it change the meaning of the code if we remove the $s (as a plan B solution)?

<img width="978" alt="Bildschirmfoto 2021-12-23 um 14 17 32" src="https://user-images.githubusercontent.com/28007169/147250515-2d8e29ec-1fc5-462a-b4ef-212940be9b1f.png">

Reply:
Oh my! 🙉 
Yes it would change the meaning if we removed the dollar sign... we must keep it

Reply:
Another option would be include an image... BUUUUT I think that would defeat part of the purpose of the article, which is to allow users to copy/paste code.
We could possibly host this code in github externally though, it would make it a bit harder for the user though.

Reply:
Yes... both the dollar sign and the curly braces need to be kept. Maybe jekyll has a way to escape them 🤔

Reply:
> {% raw %}
Hello, my name is {{name}}.
{% endraw %}

https://stackoverflow.com/questions/24102498/escaping-double-curly-braces-inside-a-markdown-code-block-in-jekyll

Reply:
I added it to the YAML file, let's see if it works.